linera-execution/src/block_executor.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{sync::Arc, vec};
+
+use custom_debug_derive::Debug;
+use linera_base::{
+    crypto::BcsHashable,
+    data_types::{Amount, Blob, Event, OracleResponse, Timestamp},
+    doc_scalar, hex_debug,
+    identifiers::{AccountOwner, BlobType, ChainId, ChannelFullName, Destination, Owner},
+};
+use linera_views::{context::Context, views::View};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    resources::Sources, ExecutionError, ExecutionRuntimeContext, ExecutionStateView, Message,
+    MessageContext, MessageKind, Operation, OperationContext, OutgoingMessage, ResourceController,
+    ResourceTracker, SystemMessage, TransactionTracker,
+};
+
+/// The execution result of a single operation.
+#[derive(Debug, Default, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub struct OperationResult(
+    #[debug(with = "hex_debug")]
+    #[serde(with = "serde_bytes")]
+    pub Vec<u8>,
+);
+
+impl<'de> BcsHashable<'de> for OperationResult {}
+
+doc_scalar!(
+    OperationResult,
+    "The execution result of a single operation."
+);
+
+pub struct BlockExecutor<'a, C> {
+    pub state: &'a mut ExecutionStateView<C>,
+    pub resource_controller: ResourceController<Option<Owner>>,
+    replaying_oracle_responses: Option<vec::IntoIter<Vec<OracleResponse>>>,
+    next_message_index: u32,
+    next_application_index: u32,
+    local_time: Timestamp,
+    pub oracle_responses: Vec<Vec<OracleResponse>>,
+    pub events: Vec<Vec<Event>>,
+    pub blobs: Vec<Vec<Blob>>,
+    pub messages: Vec<Vec<OutgoingMessage>>,
+    pub operation_results: Vec<OperationResult>,
+    pub subscribe: Vec<(ChannelFullName, ChainId)>,
+    pub unsubscribe: Vec<(ChannelFullName, ChainId)>,
+}
+
+impl<'a, C> BlockExecutor<'a, C>
+where
+    C: Context + Clone + Send + Sync + 'static,
+    C::Extra: ExecutionRuntimeContext,
+{
+    pub fn new(
+        state: &'a mut ExecutionStateView<C>,
+        account: Option<Owner>,
+        replaying_oracle_responses: Option<Vec<Vec<OracleResponse>>>,
+        local_time: Timestamp,
+    ) -> Result<Self, ExecutionError> {
+        let (_, committee) = state
+            .system
+            .current_committee()
+            .ok_or_else(|| ExecutionError::InactiveChain)?;
```

Comment:
I we don't merge `new` and `track_block` I'd extract this out and make `new` non-fallible. Nothing else here can fail and we won't need to add `with_execution_context` on the call site.